### PR TITLE
ENH: special: computing derivatives by simple autodifferentiation

### PR DESCRIPTION
**Pull request by izaid**
This PR proposes a clean-up of the C / C++ part of `special` that handles derivatives. It follows on from experiences with https://github.com/scipy/scipy/pull/20539

There are many functions in `special` that compute derivatives. This, to me, is very valuable and it should be maintained. However, during work on the above PR, it became clear that a lot of the functions in `special` that compute derivatives do so just by manual differentiation throughout the code. This means all expressions have been differentiated by hand, and indeed I did that myself to add second-order differentiation to `legendre_p`, `assoc_legendre_p`, `sph_legendre_p`, and `sph_harm`. It was a lot of work, it looks quite awkward in the code (see all the usages of `std::tuple`), and it's not easy to verify correctness with all these analytic expressions that have been inserted.

This PR is intended to sort that out, with an eye towards easily enabling derivatives (where wanted) throughout the rest of `special`. It only touches C / C++ code, nothing changes at the Python level.

What I have done is added a very simple C++ implementation of forward autodifferentation. Note this is not _backward_ autodifferentiation, there is no dynamic expression graph or anything like that. Instead we rely on dual numbers, which I have implemented as a class `dual<T, Orders...>` where `T` is the scalar type and `Orders` is just the derivative orders. A dual number can be thought of as something similar to a complex number, where instead of `i` with `i^2 = -1` we have `eps` with `eps^(N+1) = 0`. Here `N` is the maximum order of the derivative you are interested in. The storage is literally just the derivative values, there is nothing extra. So if I want to compute `f(x), f'(x), f''(x)`, the storage is indeed just those 3 values.

It also works in the multivariate case, by (mathematically) introducing an `eps` for each variable. Programmatically, this corresponds to the natural table of partial derivatives of `f(x1, ..., xn)` with respect to `x1, ... xn`. So, yes, for second order derivatives of two variables the storage is just `d^i d^j f(x, y) / dx^i dy^j` for `0 <= i, j <= 2`.

How does this work? It works by operator overloading and providing the appropriate definitions of functions, that's all. I've added what I needed for this PR, which includes arithmetic operators `+, -, *, /` and some basic functions like `sin, cos, exp, sqrt`. Aside from removing a lot of unnecessary code (great!), there were only the following very minor changes that are needed to make use of this. Suppose you have a function `T f(T x, ...)` and you want the derivative. All you need to ensure is the following:

1. Every function that `f` calls internally has an overload for `dual<T, Orders...>`. All operators work by operator overloading, so really this just means we need to be aware of simple functions like `sin, cos, log, exp, sqrt, ...` that are defined in `cmath`. This means we can't call `std::sin` directly, but instead now import `std::sin` into the `xsf` namespace by `using std::sin` and add an appropriate definition for `dual<T, Orders...> sin(dual<T, Orders...> x)`. See `dual.h` for how that works, with the imports in `config.h`.

2. There was a slightly issue with functions like `std::complex<T> f(T z, ...)` that go from (say) `double` to `std::complex<double>`. The problem with this is that it would try to instantiate `std::complex<dual<double, Orders...>>` when what we really want is `dual<std::complex<double>, Orders...>`. Mathematically, both are actually fine, and indeed I made the former work too. Programatically, we shouldn't use the former as it means the complex real and imaginary parts are no longer beside each in memory (weird, conflicts with NumPy) and `std::complex<T>` cannot be instantiated for any user type according to the C++ standard (this worked on my compiler for testing, but not on CI). The answer is to define a simple alias in `xsf` called `complex<T>` that maps `complex<dual<T>> -> dual<complex<T>>` otherwise is just equal to `std::complex`.

If you look at the functions in legendre.h, you'll see they are all vastly simpler without the explicit derivative handling and don't really look much different with the changes from 1. and 2. Again, this is all C / C++ changes, nothing changed at the Python level.

The other great benefit of this is that `dual<T, Orders...>` is memory-compatible with NumPy arrays that store the derivatives along the last dimensions. It's totally analogous to how `std::complex<T>` is memory-compatible with complex NumPy arrays. This is a great benefit that can be made use of in ufuncs, to automatically get the derivatives just by casting appropriately before calling the function.

I'm hopeful this will help a lot in `special`, and removes a lot of maintenance burden while bringing differentiation to other functions. I don't think we need any other autodifferentiation features besides what is defined here.

### CODE REVIEW

**Comment by izaid** [comment]
A few other notes on the details.

There will be a follow-up PR to synchronize a few things in Python. When that happens, the helper function `dual_assign_grad` can be deleted and all the derivative code doesn't need to know about `std::tuple` anymore.

I also want to point out that I've adapted all the functions we dealt with in https://github.com/scipy/scipy/pull/20539 to use this. That includes `legendre_p, assoc_legendre_p, sph_legendre_p, and sph_harm_y`. This is a good range already, including real functions, complex functions with branch cuts, and multivariate functions (`sph_harm_y`).

**Comment by izaid** [comment]
cc @steppi, this is ready for review

**Comment by izaid** [comment]
Just to summarize, I did a bit more work here on simplifying the wrappers, which is done. I wanted it to be possible to go directly from C++ function to ufunc, which wasn't so possible before for functions with derivatives but is now. This PR is done, just needs review.

**Comment by ev-br** [comment]
(Not a review) Just dropped in to say this is insanely cool!

File: scipy/special/xsf/legendre.h [hunk 1]
```diff
@@ -9,38 +10,22 @@ template <typename T>
...
-        T fac0 = -T(n - 1) / T(n);
-        T fac1 = T(2 * n - 1) / T(n);
-
-        tuples::assign(res, {{fac0, fac1 * z}, {0, fac1}});
-    }
-
-    void operator()(int n, std::tuple<T (&)[2], T (&)[2], T (&)[2]> res) const {
+    void operator()(int n, T (&res)[2]) const {
         T fac0 = -T(n - 1) / T(n);
         T fac1 = T(2 * n - 1) / T(n);
```
Comment (steppi) [inline scipy/special/xsf/legendre.h:25]: As we discussed offline, it's wasteful to perform these regular scalar operations on dual numbers. You'd suggested to define something like `real_type_t` for dual numbers, so we can get the correct type here. I checked and confirmed that doing so dramatically speeds things up.

**Review by steppi — APPROVED** [review]
This is very cool @izaid. I've spend time understanding it and was surprised how simple it was to implement forward autograd in terms of dual numbers like this. The simplification in the scalar kernels is admirable. 

I've confirmed that binary size and compilation times do not increase. Runtime performance is degraded compared to the original hand differentiation, but can be brought back to parity by implementing the optimizations I mention in my other comments. I will follow up shortly with a PR for these optimizations since I know you won't have bandwidth to do it in the near future.

Nice work, and agreed that this will make it much easier to add and maintain derivatives for special functions compared to the hand differentiation employed by Zhang and Jin.

**Comment by izaid** [comment]
Hooray! This is one I'm particularly proud of.

**Comment by fancidev** [comment]
Just dropping by to say this is really amazing, beautiful work!!

**Pull request merged**